eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Utxos.hs

```diff
@@ -1,59 +1,143 @@
+  updateUTxOState,
+ )
+import Cardano.Ledger.UTxO (EraUTxO (..))
+import Control.State.Transition.Extended
+import Debug.Trace (traceEvent)
+import Lens.Micro
 
 instance
-  ( EraTxOut era
+  ( AlonzoEraTx era
+  , AlonzoEraPParams era
+  , BabbageEraTxBody era
+  , EraTxOut era
   , EraGovernance era
   , EraTxCert era
+  , EraUTxO era
+  , ExtendedUTxO era
+  , EraPlutusContext 'PlutusV1 era
+  , GovernanceState era ~ ConwayGovernance era
   , Script era ~ AlonzoScript era
+  , ScriptsNeeded era ~ AlonzoScriptsNeeded era
+  , Tx era ~ AlonzoTx era
   , Eq (PPUPPredFailure era)
   , Show (PPUPPredFailure era)
   ) =>
   STS (ConwayUTXOS era)
   where
-  type BaseM (ConwayUTXOS era) = ShelleyBase
+  type BaseM (ConwayUTXOS era) = Cardano.Ledger.BaseTypes.ShelleyBase
   type Environment (ConwayUTXOS era) = UtxoEnv era
   type State (ConwayUTXOS era) = UTxOState era
   type Signal (ConwayUTXOS era) = AlonzoTx era
   type PredicateFailure (ConwayUTXOS era) = AlonzoUtxosPredFailure era
   type Event (ConwayUTXOS era) = AlonzoUtxosEvent era
 
-  transitionRules = []
+  transitionRules = [utxosTransition]
 
 instance
-  ( EraTxOut era
+  ( AlonzoEraTx era
+  , BabbageEraTxBody era
   , EraGovernance era
+  , EraPlutusContext 'PlutusV1 era
+  , EraTxOut era
   , EraTxCert era
-  , PredicateFailure (EraRule "UTXOS" era) ~ AlonzoUtxosPredFailure era
+  , EraUTxO era
+  , ExtendedUTxO era
   , Event (EraRule "UTXOS" era) ~ AlonzoUtxosEvent era
+  , GovernanceState era ~ ConwayGovernance era
+  , PredicateFailure (EraRule "UTXOS" era) ~ AlonzoUtxosPredFailure era
+  , Script era ~ AlonzoScript era
+  , ScriptsNeeded era ~ AlonzoScriptsNeeded era
+  , Tx era ~ AlonzoTx era
   , Eq (PPUPPredFailure era)
   , Show (PPUPPredFailure era)
-  , Script era ~ AlonzoScript era
   ) =>
   Embed (ConwayUTXOS era) (BabbageUTXO era)
   where
   wrapFailed = AlonzoInBabbageUtxoPredFailure . UtxosFailure
   wrapEvent = UtxosEvent
+
+utxosTransition ::
+  forall era.
+  ( AlonzoEraTx era
+  , BabbageEraTxBody era
+  , EraGovernance era
+  , EraPlutusContext 'PlutusV1 era
+  , EraUTxO era
+  , ExtendedUTxO era
+  , GovernanceState era ~ ConwayGovernance era
+  , Script era ~ AlonzoScript era
+  , ScriptsNeeded era ~ AlonzoScriptsNeeded era
+  , Tx era ~ AlonzoTx era
+  , Eq (PPUPPredFailure era)
+  , Show (PPUPPredFailure era)
+  ) =>
+  TransitionRule (ConwayUTXOS era)
+utxosTransition =
+  judgmentContext >>= \(TRC (_, _, tx)) -> do
+    case tx ^. isValidTxL of
+      IsValid True -> scriptsYes
+      IsValid False -> scriptsNo
+
+scriptsYes ::
+  forall era.
+  ( AlonzoEraTx era
+  , EraUTxO era
+  , EraPlutusContext 'PlutusV1 era
+  , ExtendedUTxO era
+  , Script era ~ AlonzoScript era
+  , ScriptsNeeded era ~ AlonzoScriptsNeeded era
+  , Tx era ~ AlonzoTx era
+  , STS (ConwayUTXOS era)
+  ) =>
+  TransitionRule (ConwayUTXOS era)
+scriptsYes = do
+  TRC (UtxoEnv _ pp dpstate _, u@(UTxOState utxo _ _ pup _), tx) <-
+    judgmentContext
+  let txBody = body tx
```

Comment:
It is better to use lenses, this lets us avoid constraints like this: `Tx era ~ AlonzoTx era`
```suggestion
  let txBody = tx ^. bodyTxL
```